File: client/tests/skills.test.ts
/**
 * Moves that are not attacks.
 *
 * PAC records a damage figure for 300 of its 531 abilities; the rest are heals,
 * buffs and movement, and the fallback turned every one of them into "hit the
 * target, splash the crowd for half". Fourteen of 57 cards cast the identical
 * effect under different names -- Teleport as a nuke, Agility as a nuke, Wish
 * as a nuke, and Happy Hour, which doubles prize money and does nothing at all
 * in a battle, as a nuke.
 *
 * These check that a move does what the move is.
 */

import { describe, it, expect } from "vitest";
import { Match } from "../src/core/match";
import { config } from "../src/core/config";
import { spawn } from "../src/core/deploy";
import { castSkill, applyHit } from "../src/core/combat";
import { byId, build } from "../src/core/cards";
import * as evolution from "../src/core/evolution";
import { MOVE_STATUS } from "../src/core/status";
import * as skills from "../src/core/skills";

/** A caster and something to aim at, both awake. */
function pair(id: string) {
  const m = new Match(5);
  const u = spawn(m, byId(id)!, config.PLAYER, 190, 400);
  u.spawning = 0;
  const foe = spawn(m, byId("charmander")!, config.ENEMY, 190, 380);
  foe.spawning = 0;
  return { m, u, foe };
}

describe("a move does what the move is", () => {
  it("Teleport moves the caster, and toward the enemy", () => {
    const { m, u, foe } = pair("abra");
    const before = u.y;
    castSkill(m, u, foe, 1);
    // The player marches up the board, so forward is a smaller y. Getting this
    // sign wrong -- which a `side === 0` test does, since PLAYER is 1 -- sends
    // it backwards into its own half.
    expect(u.y).toBeLessThan(before);
    expect(before - u.y).toBeCloseTo(64, 0);
  });

  it("Teleport does not throw the caster off the board", () => {
    const { m, u, foe } = pair("abra");
    u.y = 10;
    castSkill(m, u, foe, 1);
    expect(u.y).toBeGreaterThanOrEqual(0);
    expect(u.y).toBeLessThanOrEqual(config.arenaHeight);
  });

  it("Wish heals, and cannot overheal", () => {
    const { m, u, foe } = pair("togepi");
    u.hp = u.maxHP * 0.5;
    castSkill(m, u, foe, 1);
    expect(u.hp).toBeGreaterThan(u.maxHP * 0.5);

    u.hp = u.maxHP;
    castSkill(m, u, foe, 1);
    expect(u.hp).toBe(u.maxHP);
  });

  it("Agility raises speed, Defense Curl raises armour instead", () => {
    const a = pair("rattata");
    const speed = a.u.speed;
    castSkill(a.m, a.u, a.foe, 1);
    expect(a.u.speed).toBeGreaterThan(speed);

    const b = pair("bronzor");
    const { speed: bSpeed, def } = { speed: b.u.speed, def: b.u.def ?? 0 };
    castSkill(b.m, b.u, b.foe, 1);
    expect(b.u.speed).toBe(bSpeed);
    expect(b.u.def ?? 0).toBeGreaterThan(def);
  });

  it("a non-damage move deals no damage", () => {
    // The whole point: Teleport used to take a chunk out of whatever it was
    // aimed at, because the fallback is damage.
    const { m, u, foe } = pair("abra");
    const hp = foe.hp;
    castSkill(m, u, foe, 1);
    expect(foe.hp).toBe(hp);
  });

  it("Happy Hour pays elixir, not damage, and respects the cap", () => {
    // PAC's prize money in our currency. There is no money here, but there is
    // a resource you spend, and that is the same thing.
    const { m, u, foe } = pair("eevee");
    m.elixir[config.PLAYER] = 4;
    const hp = foe.hp;
    castSkill(m, u, foe, 1);
    expect(m.elixir[config.PLAYER]).toBe(5);
    expect(foe.hp).toBe(hp);

    m.elixir[config.PLAYER] = config.elixirMax;
    castSkill(m, u, foe, 1);
    expect(m.elixir[config.PLAYER]).toBe(config.elixirMax);
  });

  it("a real attack still hits", () => {
    const { m, u, foe } = pair("charmander");
    const hp = foe.hp;
    castSkill(m, u, foe, 1);
    expect(foe.hp).toBeLessThan(hp);
  });
});

describe("Eevee's branches are actually different", () => {
  it("only the base form keeps Happy Hour, and it means elixir now", () => {
    // The branches used to share it, so eight creatures cast one effect.
    for (const b of evolution.branchesFor("eevee") ?? []) {
      expect(build(b).skill).not.toBe("HAPPY_HOUR");
    }
    expect(build("eevee").skill).toBe("HAPPY_HOUR");
    expect(skills.MOVE_EFFECT.HAPPY_HOUR?.kind).toBe("elixir");
  });

  it("every branch carries a different status", () => {
    const branches = evolution.branchesFor("eevee") ?? [];
    expect(branches.length).toBeGreaterThan(4);

    const kinds = branches.map((b) => MOVE_STATUS[build(b).skill]?.kind);
    for (const k of kinds) expect(k).toBeDefined();
    // Distinct, or the choice is cosmetic again.
    expect(new Set(kinds).size).toBe(kinds.length);
  });
});

describe("moves powered by something other than attack", () => {
  it("uses PAC's formula, not a generic fallback", () => {
    // All six were treated as "attack x a constant". None of them are.
    expect(skills.POWERED.IRON_TAIL.from).toBe("def");
    expect(skills.POWERED.ROLLOUT.from).toBe("def");
    expect(skills.POWERED.BODY_SLAM.from).toBe("maxHP");
    expect(skills.POWERED.FOUL_PLAY.from).toBe("targetAttack");
    expect(skills.POWERED.METEOR_MASH.from).toBe("attack");
    expect(skills.POWERED.WOOD_HAMMER.from).toBe("attack");
  });

  it("Iron Tail hits for the caster's own defence", () => {
    const { m, u, foe } = pair("onix");
    u.def = 20;
    const hard = (() => { const hp = foe.hp; castSkill(m, u, foe, 1); return hp - foe.hp; })();
    const b = pair("onix");
    b.u.def = 2;
    const soft = (() => { const hp = b.foe.hp; castSkill(b.m, b.u, b.foe, 1); return hp - b.foe.hp; })();
    expect(hard).toBeGreaterThan(soft);
  });

  it("Body Slam hits for a share of the caster's own health", () => {
    const { m, u, foe } = pair("snorlax");
    const hp = foe.hp;
    castSkill(m, u, foe, 1);
    // Snorlax is the biggest body on the roster; the move should reflect that.
    expect(hp - foe.hp).toBeGreaterThan(u.maxHP * 0.2);
  });

  it("Rollout raises its own defence, so each cast hits harder than the last", () => {
    const { m, u, foe } = pair("sandshrew");
    const first = (() => { const hp = foe.hp; castSkill(m, u, foe, 1); return hp - foe.hp; })();
    const second = (() => { const hp = foe.hp; castSkill(m, u, foe, 1); return hp - foe.hp; })();
    expect(second).toBeGreaterThan(first);   // the ramp, with no counter anywhere
  });

  it("no roster move is forced physical -- PAC calls them all special", () => {
    expect((skills as Record<string, unknown>).PHYSICAL_MOVES).toBeUndefined();
  });
});

describe("Foul Play strikes with the target's attack", () => {
  it("hits harder against a strong target than a weak one", () => {
    const hit = (targetDamage: number) => {
      const { m, u, foe } = pair("sandile");
      foe.damage = targetDamage;
      foe.def = 0; foe.speDef = 0;
      const hp = foe.hp;
      castSkill(m, u, foe, 1);
      return hp - foe.hp;
    };
    // The caster is unchanged between these two; only what it is fighting is.
    expect(hit(80)).toBeGreaterThan(hit(10));
  });

  it("falls back to its own damage against a tower, which has no attack to borrow", () => {
    const { m, u } = pair("sandile");
    const tower = m.towers.find((t) => t.side !== u.side)!;
    const hp = tower.hp;
    castSkill(m, u, tower, 1);
    expect(tower.hp).toBeLessThan(hp);
  });
});

describe("the moves left as damage were decided, not forgotten", () => {
  it("records why", () => {
    for (const m of ["TRANSFORM"]) {
      expect(skills.DELIBERATELY_DAMAGE.has(m)).toBe(true);
      expect(skills.MOVE_EFFECT[m]).toBeUndefined();
    }
  });
});

describe("Electrify shields, rather than retyping anything", () => {
  it("shields itself when there is no ally, and deals no damage", () => {
    // The normal case for a win condition: it runs ahead of everything.
    const { m, u, foe } = pair("yamper");
    const hp = foe.hp;
    castSkill(m, u, foe, 1);
    expect(u.shield).toBeGreaterThan(0);
    expect(foe.hp).toBe(hp);
  });

  it("prefers the strongest ally standing with it", () => {
    const m = new Match(5);
    const y = spawn(m, byId("yamper")!, config.PLAYER, 190, 400); y.spawning = 0;
    const tank = spawn(m, byId("snorlax")!, config.PLAYER, 195, 400); tank.spawning = 0;
    const foe = spawn(m, byId("charmander")!, config.ENEMY, 190, 380); foe.spawning = 0;

    castSkill(m, y, foe, 1);
    expect(tank.shield).toBeGreaterThan(0);
    expect(y.shield).toBe(0);
  });

  it("soaks damage before health, and lets the overflow through", () => {
    const { m, u, foe } = pair("yamper");
    castSkill(m, u, foe, 1);
    const shield = u.shield;
    const hp = u.hp;

    applyHit(m, u, 5, 1, foe);
    expect(u.hp).toBe(hp);                       // absorbed entirely
    expect(u.shield).toBeLessThan(shield);

    applyHit(m, u, 10_000, 1, foe);
    expect(u.shield).toBe(0);                    // drained, not negative
    expect(u.hp).toBeLessThan(hp);               // and the rest landed
  });

  it("a shield never becomes negative health", () => {
    const { m, u, foe } = pair("yamper");
    castSkill(m, u, foe, 1);
    applyHit(m, u, 1, 1, foe);
    expect(u.shield).toBeGreaterThanOrEqual(0);
  });
});

describe("a blink lands somewhere legal", () => {
  it("never finishes inside a tower", () => {
    // Teleport ignores everything between here and there, which is the point.
    // It must not ignore where it arrives: Abra used to land in the stonework,
    // caught by the "keeps ground units out of towers" invariant on two of six
    // matches rather than by anything here.
    const m = new Match(5);
    const foe = spawn(m, byId("charmander")!, config.ENEMY, 190, 200);
    foe.spawning = 0;

    for (const t of m.towers) {
      const u = spawn(m, byId("abra")!, config.PLAYER, t.x, t.y + 70);
      u.spawning = 0;
      castSkill(m, u, foe, 1);

      const half = config.towerSize[t.kind] / 2;
      const box = config.towerBox[t.kind];
      const dy = u.y - t.y;
      const inside = Math.abs(u.x - t.x) < half && dy > -box.up && dy < box.down;
      expect(inside).toBe(false);
    }
  });
});

File: server/src/test/java/io/github/excalibase/clashofpokemon/game/rules/SkillsTest.java
package io.github.excalibase.clashofpokemon.game.rules;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.Test;

/** The thirteen casts that are not simply damage. */
class SkillsTest {

  private static Match match() {
    Match.Options o = new Match.Options();
    o.rng = DifferentialTest.mulberry32(3);
    Match m = new Match(o);
    m.elixir.put(Side.ONE, 5.0);
    return m;
  }

  private static Unit unit(Match m, String card, Side side, double x, double y) {
    Unit u = Deploy.spawn(m, Cards.byId(card), side, x, y);
    u.spawning = 0;
    return u;
  }

  // ------------------------------------------------------------ the effects

  @Test
  void happyHourMakesElixir() {
    Match m = match();
    Unit u = unit(m, "charmander", Side.ONE, 100, 500);
    double before = m.elixir.get(Side.ONE);

    assertThat(Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("HAPPY_HOUR")))
        .isTrue();
    assertThat(m.elixir.get(Side.ONE)).isGreaterThan(before);
  }

  @Test
  void elixirNeverOverfills() {
    Match m = match();
    Unit u = unit(m, "charmander", Side.ONE, 100, 500);
    m.elixir.put(Side.ONE, (double) Rules.config().elixirMax());

    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("HAPPY_HOUR"));
    assertThat(m.elixir.get(Side.ONE)).isEqualTo(Rules.config().elixirMax());
  }

  @Test
  void teleportMovesForwardAndStaysOnTheBoard() {
    Match m = match();
    Unit u = unit(m, "abra", Side.ONE, 100, 500);
    double before = u.y;

    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("TELEPORT"));
    // Side one moves up the board, so forward is a smaller y.
    assertThat(u.y).isLessThan(before);
    assertThat(u.y).isBetween(8.0, (double) Rules.config().arenaHeight() - 8);
  }

  @Test
  void teleportWillNotLeaveTheBoard() {
    Match m = match();
    Unit u = unit(m, "abra", Side.ONE, 100, 12);
    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("TELEPORT"));
    assertThat(u.y).isGreaterThanOrEqualTo(8);
  }

  @Test
  void wishHealsEveryoneNearbyAndNoneAboveFull() {
    Match m = match();
    Unit caster = unit(m, "togepi", Side.ONE, 100, 500);
    Unit hurt = unit(m, "machop", Side.ONE, 110, 500);
    Unit whole = unit(m, "geodude", Side.ONE, 120, 500);
    hurt.hp = 1;

    Skills.applyEffect(m, m.towers, caster, List.of(caster, hurt, whole),
        Skills.effectFor("WISH"));
    assertThat(hurt.hp).isGreaterThan(1);
    assertThat(whole.hp).isEqualTo(whole.maxHP);
  }

  @Test
  void aShieldGoesOnTheToughestAllyRatherThanTheCaster() {
    // Worth most on whatever is already soaking the damage.
    Match m = match();
    Unit caster = unit(m, "pikachu", Side.ONE, 100, 500);
    Unit tank = unit(m, "geodude", Side.ONE, 110, 500);
    tank.maxHP = 9999;

    Skills.applyEffect(m, m.towers, caster, List.of(caster, tank),
        Skills.effectFor("ELECTRIFY"));
    assertThat(tank.shield).isPositive();
    assertThat(caster.shield).isZero();
  }

  @Test
  void aShieldWithNobodyElseAroundGoesOnTheCaster() {
    Match m = match();
    Unit caster = unit(m, "pikachu", Side.ONE, 100, 500);
    Skills.applyEffect(m, m.towers, caster, List.of(caster), Skills.effectFor("ELECTRIFY"));
    assertThat(caster.shield).isPositive();
  }

  @Test
  void theThreeBuffsEachTouchTheirOwnStat() {
    Match m = match();
    Unit u = unit(m, "machop", Side.ONE, 100, 500);
    double speed = u.speed;
    int def = u.def;
    int speDef = u.speDef;

    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("AGILITY"));
    assertThat(u.speed).isGreaterThan(speed);
    assertThat(u.def).isEqualTo(def);

    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("DEFENSE_CURL"));
    assertThat(u.def).isGreaterThan(def);

    Skills.applyEffect(m, m.towers, u, List.of(u), Skills.effectFor("WONDER_ROOM"));
    assertThat(u.speDef).isGreaterThan(speDef);
  }

  @Test
  void anUnknownSkillDoesNothingRatherThanSomethingWrong() {
    Match m = match();
    Unit u = unit(m, "machop", Side.ONE, 100, 500);
    assertThat(Skills.effectFor("NOT_A_MOVE")).isNull();
    assertThat(Skills.effectFor(null)).isNull();
    assertThat(Skills.applyEffect(m, m.towers, u, List.of(u), null)).isFalse();
  }

  // ----------------------------------------------------------- powered moves

  @Test
  void anOrdinaryMoveIsNotPowered() {
    Match m = match();
    Unit u = unit(m, "charmander", Side.ONE, 100, 500);
    assertThat(Skills.poweredDamage("EMBER", u, null)).isNull();
    assertThat(Skills.poweredDamage(null, u, null)).isNull();
  }

  @Test
  void bodySlamScalesWithTheCastersHealth() {
    Match m = match();
    Unit u = unit(m, "spheal", Side.ONE, 100, 500);
    Double small = Skills.poweredDamage("BODY_SLAM", u, null);
    u.maxHP *= 2;
    Double large = Skills.poweredDamage("BODY_SLAM", u, null);

    assertThat(small).isNotNull();
    assertThat(large).isGreaterThan(small);
  }

  @Test
  void foulPlayScalesWithWhateverItIsHitting() {
    Match m = match();
    Unit u = unit(m, "sneasel", Side.ONE, 100, 500);
    Unit weak = unit(m, "caterpie", Side.TWO, 100, 300);
    Unit strong = unit(m, "machop", Side.TWO, 110, 300);
    strong.damage = weak.damage * 4 + 1;

    assertThat(Skills.poweredDamage("FOUL_PLAY", u, strong))
        .isGreaterThan(Skills.poweredDamage("FOUL_PLAY", u, weak));
    // No target at all: it falls back to the caster rather than to zero.
    assertThat(Skills.poweredDamage("FOUL_PLAY", u, null)).isPositive();
  }

  @Test
  void rolloutMakesTheCasterToughenAsItGoes() {
    // The one powered move with a real side effect, which is why it cannot be
    // resolved ahead of the fight.
    Match m = match();
    Unit u = unit(m, "geodude", Side.ONE, 100, 500);
    int def = u.def;

    Double first = Skills.poweredDamage("ROLLOUT", u, null);
    assertThat(u.def).isGreaterThan(def);
    assertThat(Skills.poweredDamage("ROLLOUT", u, null)).isGreaterThan(first);
  }

  @Test
  void aPoweredMoveReadsTheStageItIsCastAt() {
    Match m = match();
    Unit early = unit(m, "machop", Side.ONE, 100, 500);
    Unit late = unit(m, "machop", Side.ONE, 110, 500);
    late.card = Cards.byId("machamp");

    assertThat(Skills.poweredDamage("METEOR_MASH", late, null))
        .isGreaterThan(Skills.poweredDamage("METEOR_MASH", early, null));
  }
}
